torch/distributed/distributed_c10d.py

```diff
@@ -276,6 +276,8 @@ def register_backend(cls, name, func, extended_api=False, devices: Optional[Unio
             Backend.backend_capability[name.lower()] = devices
 
         Backend._plugins[name.upper()] = Backend._BackendPlugin(func, extended_api)
+        # Add torch.distributed.is_custombackend_available() functional
+        setattr(torch.distributed, f'is_{name.lower()}_available', lambda : True)
```

Comment:
I'm not sure if it is a good idea to change the attributes of `torch.distributed` dynamically as that introduces more chance for conflict. Imagine the case where user imports torch.distributed, defines their own custom function to distributed, then calls `register_backend` and their custom function is overidden.

Wouldn't it make more sense for users to do this on the construction of their own custom backend? 

For example in python constructor for their backend they should explicitly add the `is_X_available`. Or if defining in C++ they should modify pybind like this https://github.com/H-Huang/torch_collective_extension/blob/main/custom_backend/include/dummy.hpp#L98-L101
to also add their own `is_available` property

Reply:
> For example in python constructor for their backend they should explicitly add the is_X_available. Or if defining in C++ they should modify pybind like this https://github.com/H-Huang/torch_collective_extension/blob/main/custom_backend/include/dummy.hpp#L98-L101
to also add their own is_available property

As you said, it is a natural choice for users to add their own is_available when they create a backend. We think it might be a possible way for the user to use `is_X_available` in the same way as the native, for example by dynamically adding "torch.distributed.is_dummpy_available()" function.  This is why we want to dynamically add the `is_X_available` to `torch.distributed` in `register_backend`.

> Or we could add an Is_available(backend) function, that checks for the backend.

Providing a public function is indeed another good approach. We have implemented an `is_backend_available` in https://github.com/pytorch/pytorch/pull/101945  that supports both built-in backends and third-party backends.

Reply:
I remove the `is_X_available` in this PR and the other https://github.com/pytorch/pytorch/pull/101945 implemented `is_backend_available` for bosh build-in backends and  third-party backends.